kad-dht/README.md

```diff
@@ -0,0 +1,288 @@
+           with the peer who sent it in `Pb`.
+		2. Otherwise, we resolve the conflict by calling `Validator.Select(best,
+           new)`:
+			1. If the new value wins, store it in `best`, and mark all formerly
+               “best" peers (`Pb`) as _outdated peers_ (`Po`). The current peer
+               becomes the new best peer (`Pb`).
+			2. If the new value loses, we add the current peer to `Po`.
+	2. If successful without a value, the response will contain the closest
+       nodes the peer knows to the key `K`. Add them to the candidate list `Pn`,
+       except for those that have already been queried.
+	3. If an error or timeout occurs, discard it.
+4. Go to 1.
+
+## Entry validation
+
+When constructing a DHT node, it is possible to supply a record `Validator`
+object conforming to this interface:
+
+``` // Validator is an interface that should be implemented by record
+validators. type Validator interface {
+
+	// Validate validates the given record, returning an error if it's
+	// invalid (e.g., expired, signed by the wrong key, etc.).
+	Validate(key string, value []byte) error
+
+	// Select selects the best record from the set of records (e.g., the
+	// newest).
+	//
+	// Decisions made by select should be stable.
+	Select(key string, values [][]byte) (int, error)
+}
+```
+
+`Validate()` is a pure function that reports the validity of a record. It may
+validate a cryptographic signature, or else. It is called on two occasions:
+
+1. To validate incoming values in response to `GET_VALUE` calls.
+2. To validate outgoing values before storing them in the network via
+   `PUT_VALUE` calls.
+
+Similarly, `Select()` is a pure function that returns the best record out of 2
+or more candidates. It may use a sequence number, a timestamp, or other
+heuristic to make the decision.
+
+## Public key records
+
+Apart from storing arbitrary values, the libp2p Kad DHT stores node public keys
+in records under the `/pk` namespace. That is, the entry `/pk/<peerID>` will
+store the public key of peer `peerID`.
+
+DHT implementations may optimise public key lookups by providing a
+`GetPublicKey(peer.ID) (ci.PubKey)` method, that, for example, first checks if
+the key exists in the local peerstore.
+
+The lookup for public key entries is identical to a standard entry lookup,
+except that a custom `Validator` strategy is applied. It checks that equality
+`SHA256(value) == peerID` stands when:
+
+1. Receiving a response from a `GET_VALUE` lookup.
+2. Storing a public key in the DHT via `PUT_VALUE`.
+
+The record is rejected if the validation fails.
+
+## Provider records
+
+Nodes must keep track of which nodes advertise that they provide a given key
+(CID). These provider advertisements should expire, by default, after 24 hours.
+These records are managed through the `ADD_PROVIDER` and `GET_PROVIDERS`
+messages.
+
+_WIP (jhiesey): explain what actually happens when `Provide()` is called._
```

Comment:
Would you like me to add this section?

Reply:
Yes, please ;-) Everywhere that's marked `jhiesey` are areas I left for you ;-)

Reply:
OK I think I've addressed them all :)